### PR TITLE
[Feature](bangc-ops): add three_interpolate_backward op.

### PR DESCRIPTION
**Pull request by ArtIntAI**
Thanks for your contribution and we appreciate it a lot. 

## 1. Motivation

add new op three_interpolate_backward

## 2. Modification

mlu-ops/bangc-ops/kernels/three_interpolate_backward/*
mlu-ops/bangc-ops/mlu_op.h
mlu-ops/bangc-ops/mlu_op_kernel.h
mlu-ops/bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/three_interpolate_backward/*
mlu-ops/docs/bangc-docs/user_guide/9_operators/*
mlu-ops/docs/bangc-docs/design_docs/three_interpolate_backward/*

## 3. Test Report

If you want to know how to do operator testing, you can see [GTest-User-Guide-zh](../docs/GTest-User-Guide-zh.md).

#### 3.1.1 Accuracy Acceptance Standard

For static threshold standard details, see: [MLU-OPS Accuracy Acceptance Standard](../docs/MLU-OPS-Accuracy-Acceptance-Standard.md).

- [x] diff1, diff1 <= 3e-3
- [x] diff2, diff2 <= 3e-3

#### 3.1.2 Operator Scheme checklist

|     No.      |           Details            |      Check Results       |
|----------------|---------------------------|---------------------|
|        1       |          Supported hardware         | MLU270 <br> MLU290 <br>MLU370 <br>MLU590|
|        2       |          Job types          |   U1    |
|        3       |         Layouts            | ARRAY     |
|        4       |         Whether multi-dimensions are supported              |      NO          |
|        5       |          Whether element zero is supported             |        YES       |
|        6       |         Data type(half/float)       |         half / float           |
|        7      |        Whether there is size limit           |             |

#### 3.1.3 New Feature Test

If you have checked the following items, please tick the relevant box.

- [x] Data type test
- [ ] Multidimensional tensor test
- [x] Layout test
- [x] Different size/integer remainder end segment/alignment misalignment test
- [x] Zero dimensional tensor test/zero element test
- [x] stability test
- [x] Multiple platform test
- [x] Gen_case module test
- [x] Nan/INF tests 
- [ ] Bug fix tests
- [x] For memory leak check details, see[GTest-User-Guide-zh](../docs/GTest-User-Guide-zh.md).
- [x] For code coverage check details, see: [GTest-User-Guide-zh](../docs/GTest-User-Guide-zh.md).
- [x] For I/O calculation efficiency check details, see: [MLU-OPS Performance Acceptance Standard](../docs/MLU-OPS-Performance-Acceptance-Standard.md).

#### 3.1.4 Parameter Check

When a new operator is submitted, the test points are given and the test results are stated.

| Test Point         | Acceptance Standard | Test Result (Error Message) |
| -------------- | -------- | -------------------- |
| Whether it conforms to the operator restriction | Normal error |                      |
| Whether illegal parameters are passed  | Normal error |                      |

### 3.2 Performance Test

See [MLU-OPS Performance Acceptance Standard](../docs/MLU-OPS-Performance-Acceptance-Standard.md) for details.

Platform ：MLU270

|Operator|Mlu_hardware_time(us)|Mlu_interface_time(us)|Mlu_io_efficiency|Mlu_compute_efficiency|Mlu_workwpace_size(Bytes)|Data_type|Shape|
|-----|----|----|----|----|----|------|-----|
|op_name|   |    |     |    |    |    |     |
|op_name|   |    |     |    |    |    |     |

Platform ：MLU290

|Operator|Mlu_hardware_time(us)|Mlu_interface_time(us)|Mlu_io_efficiency|Mlu_compute_efficiency|Mlu_workwpace_size(Bytes)|Data_type|Shape|
|-----|----|----|----|----|----|------|-----|
|op_name|   |    |     |    |    |    |     |
|op_name|   |    |     |    |    |    |     |

Platform：MLU370

|Operator|Mlu_hardware_time(us)|Mlu_interface_time(us)|Mlu_io_efficiency|Mlu_compute_efficiency|Mlu_workwpace_size(Bytes)|Data_type|Shape|
|-----|----|----|----|----|----|------|-----|
|op_name|   |    |     |    |    |    |     |
|op_name|   |    |     |    |    |    |     |

### 3.3 Summary Analysis

Please give a brief overview here, if you want to note and summarize the content.

### CODE REVIEW

File: docs/bangc-docs/design_docs/three_interpolate_backward/three_interpolate_backward.md [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+mluOpStatus_t MLUOP_WIN_API mluOpThreeInterpolateBackward(mluOpHandle_t handle,
+                                                         const mluOpTensorDescriptor_t grad_output_desc,
+                                                         const void *grad_output,
+                                                         const mluOpTensorDescriptor_t indices_desc,
+                                                         const void *indices,
+                                                         const mluOpTensorDescriptor_t weights_desc,
+                                                         const void *weights,
+                                                         const mluOpTensorDescriptor_t grad_features_desc,
+                                                         void *grad_features);
+```
```
Comment (tudejiang79) [inline docs/bangc-docs/design_docs/three_interpolate_backward/three_interpolate_backward.md:274]: 格式对齐

File: bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/three_interpolate_backward/three_interpolate_backward.h [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+namespace mluoptest {
+
+class ThreeInterpolateBackwardExecutor : public Executor {
+ public:
+  ThreeInterpolateBackwardExecutor() {}
+  ~ThreeInterpolateBackwardExecutor() {}
+
+  void paramCheck();
+  void compute();
+  void cpuCompute();
```
Comment (tudejiang79) [inline bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/three_interpolate_backward/three_interpolate_backward.h:38]:
```suggestion
  void paramCheck();override
  void compute();override
  void cpuCompute();override
```

File: docs/bangc-docs/design_docs/three_interpolate_backward/three_interpolate_backward.md (hunk 1 above)
Comment (ArtIntAI) [reply docs/bangc-docs/design_docs/three_interpolate_backward/three_interpolate_backward.md:274]: 我这看到的格式是对齐的啊

File: bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/three_interpolate_backward/three_interpolate_backward.h (hunk 2 above)
Comment (ArtIntAI) [reply bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/three_interpolate_backward/three_interpolate_backward.h]: 已经修改。

File: docs/bangc-docs/design_docs/three_interpolate_backward/three_interpolate_backward.md (hunk 1 above)
Comment (ArtIntAI) [reply docs/bangc-docs/design_docs/three_interpolate_backward/three_interpolate_backward.md:274]: 已经修改。

File: bangc-ops/mlu_op.h [hunk 3]
```diff
@@ -3493,6 +3493,93 @@ mluOpThreeInterpolateForward(mluOpHandle_t handle,
                              const mluOpTensorDescriptor_t output_desc,
                              void *output);
 
+// Group: ThreeInterpolate
+/*!
+ * @brief Computes the gradients of feature map \b grad_features based on the
+ *  inputs \b grad_output , \b indices and \b weights to perform the backpropagation
+ *  of the ::mluOpThreeInterpolateForward operator.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3500]: 所有的算子，全部使用operation，这里operator 》 operation。

File: bangc-ops/mlu_op.h [hunk 4]
```diff
@@ -3493,6 +3493,93 @@ mluOpThreeInterpolateForward(mluOpHandle_t handle,
                              const mluOpTensorDescriptor_t output_desc,
                              void *output);
 
+// Group: ThreeInterpolate
+/*!
+ * @brief Computes the gradients of feature map \b grad_features based on the
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3498]: 这里feature和features是不是多了一个？

File: bangc-ops/mlu_op.h [hunk 5]
```diff
@@ -3493,6 +3493,93 @@ mluOpThreeInterpolateForward(mluOpHandle_t handle,
...
+ * Pointer to the MLU memory that stores the gradients of features tensor. The
+ * grad_features' shape (B, C, M), B is batch size, C is channel size, M is number
+ * of elements in one input channel.
+ *
+ * @par Return
+ * - ::MLUOP_STATUS_SUCCESS, ::MLUOP_STATUS_BAD_PARAM,
+ *   ::MLUOP_STATUS_NOT_SUPPORTED
+ *
+ * @par Data Type
+ * - Data type of grad_output tensor, weights tensor and grad_features tensor should be the same.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3540]: Data types

File: bangc-ops/mlu_op.h [hunk 6]
```diff
@@ -3493,6 +3493,93 @@ mluOpThreeInterpolateForward(mluOpHandle_t handle,
...
+ * - The supported data types of input and output tensors are as follows:
+ *   - grad_output tensor: half, float.
+ *   - indices tensor: int.
+ *   - weights tensor: half, float.
+ *   - grad_features tensor: half, float.
+ *
+ *  @par Data Layout
+ *  - The supported data layout of \b grad_output, \b indices, \b weights, \b grad_features are
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3548]: data layputs

**Review by GraceGuanH — COMMENTED** [review]
Please check the comments.

File: bangc-ops/mlu_op.h (hunk 4 above)
Comment (ArtIntAI) [reply bangc-ops/mlu_op.h:3498]: gradients of feature map就是grad_features的含义

File: bangc-ops/mlu_op.h (hunk 5 above)
Comment (ArtIntAI) [reply bangc-ops/mlu_op.h]: 已经修改

File: bangc-ops/mlu_op.h (hunk 3 above)
Comment (ArtIntAI) [reply bangc-ops/mlu_op.h]: 已经修改

File: bangc-ops/mlu_op.h (hunk 4 above)
Comment (GraceGuanH) [reply bangc-ops/mlu_op.h:3498]: 忽略，看错了

**Review by GraceGuanH — APPROVED** [review]

**Review by tudejiang79 — APPROVED** [review]

File: bangc-ops/kernels/three_interpolate_backward/three_interpolate_backward.cpp [hunk 7]
```diff
@@ -0,0 +1,340 @@
...
+
+#define WEIGHT_N_LIMIT_SIZE 6
+#define INDEX_N_LIMIT_SIZE 6
+#define INPUT_NC_LIMIT_SIZE 2
+#define OUTPUT_MC_LIMIT_SIZE 2
+#define INDEX_TYPE_CONVERT_N_LIMIT_SIZE 3
+#define REPEATS_TIME_GAP 1
+#define REPEATS_TIME_MOD 0
+
+void PolicyFuncThreeInterpolateBackward(
```
Comment (tudejiang79) [inline bangc-ops/kernels/three_interpolate_backward/three_interpolate_backward.cpp:43]:
```suggestion
static void PolicyFuncThreeInterpolateBackward(
```

File: bangc-ops/kernels/three_interpolate_backward/three_interpolate_backward.cpp [hunk 8]
```diff
@@ -0,0 +1,340 @@
...
+    const mluOpHandle_t &handle, const mluOpTensorDescriptor_t &desc, int b,
+    int c, int m, int n, cnrtDim3_t *k_dim, cnrtFunctionType_t *k_type,
+    int &c_limit_size, int &m_limit_size, int &n_limit_size) {
```
Comment (tudejiang79) [inline bangc-ops/kernels/three_interpolate_backward/three_interpolate_backward.cpp:46]: 不需要返回的形参加const ；引用类型全用指针

File: bangc-ops/kernels/three_interpolate_backward/three_interpolate_backward.cpp (hunk 7 above)
Comment (ArtIntAI) [reply bangc-ops/kernels/three_interpolate_backward/three_interpolate_backward.cpp]: 已经修改。

File: bangc-ops/kernels/three_interpolate_backward/three_interpolate_backward.cpp (hunk 8 above)
Comment (ArtIntAI) [reply bangc-ops/kernels/three_interpolate_backward/three_interpolate_backward.cpp]: 已经修改。

File: bangc-ops/mlu_op_kernel.h [hunk 9]
```diff
@@ -283,6 +283,19 @@ void MLUOP_WIN_API mluOpUnionKernelThreeInterpolateForwardHalf(
     const int c, const int m, const int n, const int c_limit_size,
     const int m_limit_size, const int n_limit_size, void *output);
 
+void MLUOP_WIN_API mluOpUnionKernelThreeInterpolateBackwardFloat(
+    cnrtDim3_t k_dim, cnrtFunctionType_t k_type, cnrtQueue_t queue,
+    const void *grad_output, const void *indices, const void *weights,
+    const int b, const int c, const int m, const int n,
+    const int c_limit_size, const int m_limit_size, const int n_limit_size,
+    void *grad_features);
+void MLUOP_WIN_API mluOpUnionKernelThreeInterpolateBackwardHalf(
```
Comment (WangYuan-cn) [inline bangc-ops/mlu_op_kernel.h:292]:
```suggestion

void MLUOP_WIN_API mluOpUnionKernelThreeInterpolateBackwardHalf(
```

Comment (WangYuan-cn) [inline bangc-ops/mlu_op_kernel.h:292]:
```suggestion
    void *grad_features);

void MLUOP_WIN_API mluOpUnionKernelThreeInterpolateBackwardHalf(
```

**Review by WangYuan-cn — APPROVED** [review]

File: bangc-ops/mlu_op_kernel.h (hunk 9 above)
Comment (ArtIntAI) [reply bangc-ops/mlu_op_kernel.h]: 已经修改

File: bangc-ops/kernels/three_interpolate_backward/three_interpolate_backward.cpp [hunk 10]
```diff
@@ -0,0 +1,341 @@
+/*************************************************************************
+ * Copyright (C) [2022] by Cambricon, Inc.
```
Comment (GuoZhiBin2014) [inline bangc-ops/kernels/three_interpolate_backward/three_interpolate_backward.cpp:2]: `[2023]`

File: bangc-ops/kernels/three_interpolate_backward/three_interpolate_backward.cpp [hunk 11]
```diff
@@ -0,0 +1,341 @@
...
+  if ((mluOpGetTensorElementNum(grad_output_desc) >= LARGE_TENSOR_NUM) ||
+      (mluOpGetTensorElementNum(indices_desc) >= LARGE_TENSOR_NUM) ||
+      (mluOpGetTensorElementNum(weights_desc) >= LARGE_TENSOR_NUM) ||
+      (mluOpGetTensorElementNum(grad_features_desc) >= LARGE_TENSOR_NUM)) {
+    LOG(ERROR) << op_name << " Overflow max tensor num."
+               << " Currently, MLU-OPS supports tensor num smaller than 2^31.";
+    return MLUOP_STATUS_NOT_SUPPORTED;
+  }
+  // check zero element
+  if ((mluOpGetTensorElementNum(grad_output_desc) == 0) ||
```
Comment (GuoZhiBin2014) [inline bangc-ops/kernels/three_interpolate_backward/three_interpolate_backward.cpp:270]: 0元素是否应该直接返回success，而不是直接参数检查报错．是否有０元素场景

File: bangc-ops/mlu_op.h [hunk 12]
```diff
@@ -3493,6 +3493,93 @@ mluOpThreeInterpolateForward(mluOpHandle_t handle,
...
+ * @param[in] grad_output
+ * Pointer to the MLU memory that stores the gradients of output tensor. The grad_output's
+ * shape (B, C, N), B is batch size, C is channel size, N is the number of
+ * elements in one output channel.
+ * @param[in] indices_desc
+ * The descriptor of the indices tensors. For detailed information, see
+ * ::mluOpTensorDescriptor_t.
+ * @param[in] indices
+ * Pointer to the MLU memory that stores the input indicies tensor. The indices'
+ * shape (B, N, 3), B is batch size, C is channel size, N is the number of
```
Comment (GuoZhiBin2014) [inline bangc-ops/mlu_op.h:3518]: 删除`C is channel size,`

File: bangc-ops/kernels/three_interpolate_backward/three_interpolate_backward.cpp (hunk 11 above)
Comment (ArtIntAI) [reply bangc-ops/kernels/three_interpolate_backward/three_interpolate_backward.cpp:270]: mluops这里都是这样处理的

File: bangc-ops/kernels/three_interpolate_backward/three_interpolate_backward.cpp (hunk 10 above)
Comment (ArtIntAI) [reply bangc-ops/kernels/three_interpolate_backward/three_interpolate_backward.cpp:2]: 貌似当前都是2022,后续等统计刷新下吧

File: bangc-ops/mlu_op.h (hunk 12 above)
Comment (ArtIntAI) [reply bangc-ops/mlu_op.h:3518]: 已经在新的pr上修改

File: bangc-ops/kernels/three_interpolate_backward/three_interpolate_backward.cpp (hunk 11 above)
Comment (ArtIntAI) [reply bangc-ops/kernels/three_interpolate_backward/three_interpolate_backward.cpp:270]: 已经在新pr上修改成M支持零元素的形式

**Pull request merged**